File: javatests/google/registry/ui/js/registrar/whois_settings_test.js
// Copyright 2017 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.setTestOnly();

goog.require('goog.dispose');
goog.require('goog.dom');
goog.require('goog.dom.classlist');
goog.require('goog.testing.MockControl');
goog.require('goog.testing.PropertyReplacer');
goog.require('goog.testing.asserts');
goog.require('goog.testing.jsunit');
goog.require('goog.testing.net.XhrIo');
goog.require('registry.registrar.ConsoleTestUtil');
goog.require('registry.testing');
goog.require('registry.util');


const $ = goog.dom.getRequiredElement;
const $$ = goog.dom.getRequiredElementByClass;
const stubs = new goog.testing.PropertyReplacer();

const test = {
  testXsrfToken: '༼༎෴ ༎༽',
  testClientId: 'testClientId',
  mockControl: new goog.testing.MockControl()
};


function setUp() {
  registry.testing.addToDocument('<div id="test"/>');
  registry.testing.addToDocument('<div class="kd-butterbar"/>');
  registry.registrar.ConsoleTestUtil.renderConsoleMain($('test'), {
    xsrfToken: test.testXsrfToken,
    clientId: test.testClientId,
  });
  stubs.setPath('goog.net.XhrIo', goog.testing.net.XhrIo);
  registry.registrar.ConsoleTestUtil.setup(test);
}


function tearDown() {
  goog.dispose(test.console);
  stubs.reset();
  goog.testing.net.XhrIo.cleanup();
  test.mockControl.$tearDown();
}


/**
 * Creates test registrar.
 * @return {!Object}
 */
function createTestRegistrar() {
  return {
    emailAddress: 'test2.ui@example.com',
    clientIdentifier: 'theRegistrar',
    ianaIdentifier: 1,
    icannReferralEmail: 'lol@sloth.test',
    whoisServer: 'foo.bar.baz',
    url: 'blah.blar',
    phoneNumber: '+1.2125650000',
    faxNumber: '+1.2125650001',
    localizedAddress: {
      street: ['111 Eighth Avenue', 'Eleventh Floor', 'lol'],
      city: 'New York',
      state: 'NY',
      zip: '10011',
      countryCode: 'US'
    }};
}


function testView() {
  registry.registrar.ConsoleTestUtil.visit(test, {
    path: 'whois-settings',
    xsrfToken: test.testXsrfToken,
    clientId: test.testClientId
  });
  const testRegistrar = createTestRegistrar();
  registry.testing.assertReqMockRsp(
      test.testXsrfToken,
      '/registrar-settings',
      {op: 'read', id: 'testClientId', args: {}},
      {
        status: 'SUCCESS',
        message: 'OK',
        results: [testRegistrar]
      });
  const parsed = registry.util.parseForm('item');
  parsed.ianaIdentifier = parseInt(parsed.ianaIdentifier);
  registry.testing.assertObjectEqualsPretty(testRegistrar, parsed);
}


function testEdit() {
  testView();
  registry.testing.click($('reg-app-btn-edit'));
  $('emailAddress').value = 'test2.ui@example.com';
  $('localizedAddress.street[0]').value = 'look at me i am';
  $('localizedAddress.street[1]').value = 'the mistress of the night';
  $('localizedAddress.street[2]').value = '';
  const parsed = registry.util.parseForm('item');
  parsed.readonly = false;
  registry.testing.click($('reg-app-btn-save'));
  registry.testing.assertReqMockRsp(
      test.testXsrfToken,
      '/registrar-settings',
      {op: 'update', id: 'testClientId', args: parsed},
      {
        status: 'SUCCESS',
        message: 'OK',
        results: [parsed]
      });
}


function testEditFieldError_insertsError() {
  testView();
  registry.testing.click($('reg-app-btn-edit'));
  $('phoneNumber').value = 'foo';
  const parsed = registry.util.parseForm('item');
  parsed.readonly = false;
  registry.testing.click($('reg-app-btn-save'));
  const errMsg = 'Carpe brunchus. --Pablo';
  registry.testing.assertReqMockRsp(
      test.testXsrfToken,
      '/registrar-settings',
      {op: 'update', id: 'testClientId', args: parsed},
      {
        status: 'ERROR',
        field: 'phoneNumber',
        message: errMsg
      });
  const msgBox = goog.dom.getNextElementSibling($('phoneNumber'));
  assertTrue(goog.dom.classlist.contains(msgBox, 'kd-errormessage'));
  assertTrue(goog.dom.classlist.contains($('phoneNumber'), 'kd-formerror'));
  assertEquals(errMsg, goog.dom.getTextContent(msgBox));
}


function testEditNonFieldError_showsButterBar() {
  testView();
  registry.testing.click($('reg-app-btn-edit'));
  const parsed = registry.util.parseForm('item');
  parsed.readonly = false;
  registry.testing.click($('reg-app-btn-save'));
  const errMsg = 'One must still have chaos in oneself to be able to give ' +
      'birth to a dancing star. --Nietzsche';
  registry.testing.assertReqMockRsp(
      test.testXsrfToken,
      '/registrar-settings',
      {op: 'update', id: 'testClientId', args: parsed},
      {
        status: 'ERROR',
        message: errMsg
      });
  assertEquals(errMsg, goog.dom.getTextContent($$('kd-butterbar-text')));
}

File: core/src/test/java/google/registry/ui/server/registrar/WhoisSettingsTest.java
// Copyright 2017 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package google.registry.ui.server.registrar;

import static com.google.common.base.Strings.repeat;
import static com.google.common.truth.Truth.assertThat;
import static google.registry.testing.DatastoreHelper.loadRegistrar;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import google.registry.model.registrar.Registrar;
import google.registry.model.registrar.RegistrarAddress;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * Unit tests for security_settings.js use of {@link RegistrarSettingsAction}.
 *
 * <p>The default read and session validation tests are handled by the superclass.
 */
class WhoisSettingsTest extends RegistrarSettingsActionTestCase {

  @Test
  void testPost_update_success() throws Exception {
    Registrar modified =
        loadRegistrar(CLIENT_ID)
            .asBuilder()
            .setEmailAddress("hello.kitty@example.com")
            .setPhoneNumber("+1.2125650000")
            .setFaxNumber("+1.2125650001")
            .setUrl("http://acme.com/")
            .setWhoisServer("ns1.foo.bar")
            .setLocalizedAddress(
                new RegistrarAddress.Builder()
                    .setStreet(ImmutableList.of("76 Ninth Avenue", "Eleventh Floor"))
                    .setCity("New York")
                    .setState("NY")
                    .setZip("10009")
                    .setCountryCode("US")
                    .build())
            .build();
    Map<String, Object> response =
        action.handleJsonRequest(
            ImmutableMap.of("op", "update", "id", CLIENT_ID, "args", modified.toJsonMap()));
    assertThat(response.get("status")).isEqualTo("SUCCESS");
    assertThat(response.get("results")).isEqualTo(ImmutableList.of(modified.toJsonMap()));
    assertThat(loadRegistrar(CLIENT_ID)).isEqualTo(modified);
    assertMetric(CLIENT_ID, "update", "[OWNER]", "SUCCESS");
    verifyNotificationEmailsSent();
  }

  @Test
  void testPost_badUsStateCode_returnsFormFieldError() {
    Registrar modified =
        loadRegistrar(CLIENT_ID)
            .asBuilder()
            .setEmailAddress("hello.kitty@example.com")
            .setPhoneNumber("+1.2125650000")
            .setFaxNumber("+1.2125650001")
            .setLocalizedAddress(
                new RegistrarAddress.Builder()
                    .setStreet(ImmutableList.of("76 Ninth Avenue", "Eleventh Floor"))
                    .setCity("New York")
                    .setState("ZZ")
                    .setZip("10009")
                    .setCountryCode("US")
                    .build())
            .build();
    Map<String, Object> response =
        action.handleJsonRequest(
            ImmutableMap.of("op", "update", "id", CLIENT_ID, "args", modified.toJsonMap()));
    assertThat(response.get("status")).isEqualTo("ERROR");
    assertThat(response.get("field")).isEqualTo("localizedAddress.state");
    assertThat(response.get("message")).isEqualTo("Unknown US state code.");
    assertThat(loadRegistrar(CLIENT_ID)).isNotEqualTo(modified);
    assertMetric(CLIENT_ID, "update", "[OWNER]", "ERROR: FormFieldException");
  }

  @Test
  void testPost_badAddress_returnsFormFieldError() {
    Registrar modified =
        loadRegistrar(CLIENT_ID)
            .asBuilder()
            .setEmailAddress("hello.kitty@example.com")
            .setPhoneNumber("+1.2125650000")
            .setFaxNumber("+1.2125650001")
            .setLocalizedAddress(
                new RegistrarAddress.Builder()
                    .setStreet(ImmutableList.of("76 Ninth Avenue", repeat("lol", 200)))
                    .setCity("New York")
                    .setState("NY")
                    .setZip("10009")
                    .setCountryCode("US")
                    .build())
            .build();
    Map<String, Object> response =
        action.handleJsonRequest(
            ImmutableMap.of("op", "update", "id", CLIENT_ID, "args", modified.toJsonMap()));
    assertThat(response.get("status")).isEqualTo("ERROR");
    assertThat(response.get("field")).isEqualTo("localizedAddress.street[1]");
    assertThat((String) response.get("message"))
        .contains("Number of characters (600) not in range");
    assertThat(loadRegistrar(CLIENT_ID)).isNotEqualTo(modified);
    assertMetric(CLIENT_ID, "update", "[OWNER]", "ERROR: FormFieldException");
  }

  @Test
  void testPost_badWhoisServer_returnsFormFieldError() {
    Registrar modified =
        loadRegistrar(CLIENT_ID).asBuilder().setWhoisServer("tears@dry.tragical.lol").build();
    Map<String, Object> response =
        action.handleJsonRequest(
            ImmutableMap.of("op", "update", "id", CLIENT_ID, "args", modified.toJsonMap()));
    assertThat(response.get("status")).isEqualTo("ERROR");
    assertThat(response.get("field")).isEqualTo("whoisServer");
    assertThat(response.get("message")).isEqualTo("Not a valid hostname.");
    assertThat(loadRegistrar(CLIENT_ID)).isNotEqualTo(modified);
    assertMetric(CLIENT_ID, "update", "[OWNER]", "ERROR: FormFieldException");
  }
}
